src/devices/lightbulbs.ts

```diff
@@ -0,0 +1,546 @@
+            const hsl = RGBToHSL(rgb[0], rgb[1], rgb[2]);
+            const saturation = hsl[1];
+            this.platform.log.debug(
+              'Retrieved saturation status successfully: ',
+              saturation,
+            );
+            this.platform.log.debug(
+              'Retrieved saturation/hue status successfully: ',
+              this.Hue,
+            );
+            this.Saturation = saturation;
+          } else {
+            this.platform.log.debug('Retrieved status unsuccessfully.');
+            this.Saturation = this.On ? 100 : 0;
+          }
+        }
+    }
+  }
+
+  async refreshStatus() {
+    // Namespace
+    switch (this.device.model) {
+      case 'MSL-320':
+        this.Namespace = 'Appliance.System.Online';
+        break;
+      default:
+        this.Namespace = 'Appliance.System.All';
+    }
+
+    try {
+      this.platform.log.debug('%s - Reading', this.device.model, `${this.device.deviceUrl}/status`);
+      const deviceStatus = (
+        await this.platform.axios({
+          url: `http://${this.device.deviceUrl}/config`,
+          method: 'post',
+          data: {
+            payload: {},
+            header: {
+              messageId: `${this.device.messageId}`,
+              method: 'GET',
+              from: `http://${this.device.deviceUrl}/config`,
+              namespace: this.Namespace,
+              timestamp: this.device.timestamp,
+              sign: `${this.device.sign}`,
+              payloadVersion: 1,
+            },
+          },
+        },
+        )).data;
+      this.platform.log.debug(
+        '%s %s refreshStatus -',
+        this.device.model,
+        this.accessory.displayName,
+        JSON.stringify(deviceStatus),
+      );
+      this.parseStatus();
+      this.updateHomeKitCharacteristics();
+    } catch (e) {
+      this.platform.log.error(
+        '%s - Failed to refresh status of %s: %s',
+        this.device.model,
+        this.device.name,
+        JSON.stringify(e.message),
+        this.platform.log.debug('%s %s -', this.device.model, this.accessory.displayName, JSON.stringify(e)),
+      );
+      this.apiError(e);
+    }
+  }
+
+  /**
+ * Pushes the requested changes to the Meross Device
+ */
+  async pushOnChanges() {
+    this.Payload = {
+      togglex: {
+        onoff: this.On ? 1 : 0,
+        channel: `${this.device.channel}`,
+      },
+    };
+
+    // Data Info
+    this.Data = {
+      payload: this.Payload,
+      header: {
+        messageId: `${this.device.messageId}`,
+        method: 'SET',
+        from: `http://${this.device.deviceUrl}/config`,
+        namespace: 'Appliance.Control.ToggleX',
+        timestamp: this.device.timestamp,
+        sign: `${this.device.sign}`,
+        payloadVersion: 1,
+      },
+    };
+
+    // Make request
+    const push = await this.platform.axios({
+      url: `http://${this.device.deviceUrl}/config`,
+      method: 'post',
+      data: this.Data,
+    },
+    );
+    if (this.On) {
```

Comment:
What exactly does this if statement do, why is it necessary? If I remove it, my MSL-420 works fine again.

Reply:
weird because all that does is put text in the log to say either that it has been turned on or off.

Reply:
I get the following log output in the homebridge console (using the code from 4.0.1-beta4):

**Turning the bulb on (1 time):**

```
[3/4/2021, 8:07:18 PM] [Meross] Sending request On for BULB
[3/4/2021, 8:07:18 PM] [Meross] Sending request undefined for BULB
[3/4/2021, 8:07:18 PM] [Meross] Sending request On for BULB
[3/4/2021, 8:07:18 PM] [Meross] Sending request On for BULB
```

**Trying to turn it back off (1 time):**

```
[3/4/2021, 8:08:56 PM] [Meross] Sending request Off for BULB
[3/4/2021, 8:08:56 PM] [Meross] Sending request undefined for BULB
[3/4/2021, 8:08:56 PM] [Meross] Sending request Off for BULB
[3/4/2021, 8:08:56 PM] [Meross] Sending request Off for BULB
```

I don't know why, but it did turn off this time, although I did not change anything in the code this time 🤨 But it's clear that something is going, where the request is sent multiple times.

Reply:
When I select a different color, it's flashing in some other color before it's using the color I selected